### PR TITLE
Script to build the fatpack plugin versions

### PR DESCRIPTION
**Pull request by CPbN**
Hello,

Based on our discussion in #346, here is a Bash script to automatically build the fastpack version of every plugin.

Thank you 👍

### CODE REVIEW

**Comment by CPbN** [comment]
Done !

**Comment by Sims24** [comment]
Thanks for your contribution, we're testing this today and merge it if everything is ok ;)

**Comment by joschi99** [comment]
Hi @CPbN,

great work. Found a missing library (for mysql checks), you need to add dbi.pm

`cp -R --parents centreon/plugins/{dbi,misc,mode,options,output,perfdata,script,statefile,values}.pm centreon/plugins/templates/ centreon/plugins/alternative/ fatpack/build/lib/`

**Comment by CPbN** [comment]
Thank you for your feedback @joschi99 :+1: 
Mmmh with your workaround we would have `dbi.pm` in every plugin, which I think we do not want.
I think I've found where the issue comes from, I'll push a fix !

**Comment by joschi99** [comment]
👍 Great. I think the same issue we have on http.pm to check for example Apache. Another issue that we have to resolve is for example on elasticsearch:
```
./apps-elasticsearch-restapi.pl --hostname=127.0.0.1 --port=9200 --proto=http --plugin=apps::elasticsearch::restapi::plugin --mode=cluster
UNKNOWN: Cannot load module --custommode.
Can't locate apps/elasticsearch/restapi/custom/api.pm in @INC (@INC contains: /opt/bi-s/software/plugins/centreon-plugins/centreon-plugins-20180528/fatpack FatPacked::8706016=HASH(0x84d7e0) /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 .) at ./apps-elasticsearch-restapi.pl line 1696.
```

**Comment by benrubson** [comment]
I've just submitted a much more generic version, and complete solution.
Every added file is now processed recursively to find additional required `.pm` files.
It then corrects previous bugs 👍

**Comment by CPbN** [comment]
I've just submitted a much more generic version, and complete solution.
Every added file is now processed recursively to find for additional required `.pm` files.
I then corrects previous bugs 👍

File: centreon/plugins/script_sql.pm [hunk 1]
```diff
@@ -48,7 +48,7 @@ sub new {
                                   );
     $self->{version} = '1.0';
     %{$self->{modes}} = ();
-    %{$self->{sql_modes}} = ('dbi' => 'centreon::plugins::dbi');
+    $self->{sql_modes}{api} = 'centreon::plugins::dbi';
```
Comment (Sims24) [inline centreon/plugins/script_sql.pm]: cannot imagine that this change doesn't break things, did you really test this ? that is totally different

Comment (CPbN) [reply centreon/plugins/script_sql.pm]: yes there are 2 different `(custom_)?modes` pattern caught and I wrongly rewrote this line using the wrong pattern...
I mainly concentrated to include every needed file, and did some basic testing with a few plugins to be sure they launched correctly.
SQL ones do now.
Thank you for your feedback :+1:

**Comment by joschi99** [comment]
Hi @CPbN,
your fix is working very well. I can confirm that the change on `centreon/plugins/script_sql.pm` is working. 👍 

Found another issue using `dyn-mode` with mysql, but I can imagine that this problem is related probably also on other checks using dyn-mode:
```
./database-mysql.pl --plugin=database::mysql::plugin --dyn-mode=apps::centreon::sql::mode::countservices --host=127.0.0.1 --port=3306 --username=user --password=pwd --warning-host= --critical-host= --warning-service= --critical-service= --filter-poller='^Central$' --verbose
UNKNOWN: Cannot load module --dyn-mode.
Can't locate apps/centreon/sql/mode/countservices.pm in @INC (@INC contains: /opt/bi-s/software/plugins/centreon-plugins/centreon-plugins-20180528/fatpack FatPacked::15882328=HASH(0xf25858) /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 .) at ./database-mysql.pl line 1193.
```

File: centreon/plugins/script_sql.pm (hunk 1 above)
Comment (Sims24) [reply centreon/plugins/script_sql.pm]: @CPbN please make it looks as it was originally as a one line :) thanks

**Comment by CPbN** [comment]
Yes this is due to the fact that _dyn-modes_ are not referenced anywhere, well at least I did not found out.
If you launch your command from the parent directory, it will succeed :
```
./fatpack/database-mysql.pl --plugin=database::mysql::plugin --dyn-mode=apps::centreon::sql::mode::countservices
```
But of course this is not really what we want.
We must find a smart way to include _dyn-modes_.

**Comment by joschi99** [comment]
Exactly, otherwise we cannot use the Perl Connector and benefit from the performance improve. If I use the packed version from Centreon Plugin Pack it is working.

File: centreon/plugins/script_sql.pm (hunk 1 above)
Comment (CPbN) [reply centreon/plugins/script_sql.pm]: It will complicate pattern matching, but let's have a look !

**Comment by CPbN** [comment]
@Sims24 @cgagnaire, any way, going through the code of a plugin, to know that it may depend on some _dyn-modes_ files ?
I did not get further into this yet, but perhaps you already have the answer !
Thx :+1:

File: centreon/plugins/script_sql.pm (hunk 1 above)
Comment (CPbN) [reply centreon/plugins/script_sql.pm]: Done, I've found a way without complicating pattern matching :+1:

**Comment by CPbN** [comment]
OK, I found a way to support `--dyn-mode`.
I simply added the existing relations to the beginning of the script.
To maintain according to future _dyn-modes_ directories or calling plugins.
I think we should be OK now 👍

**Comment by joschi99** [comment]
Hi @CPbN,
something is wrong with you last fix. `os-linux-snmp.pl` for example is missing `cpu.pm`, `storage.pm` ecc, same issue on other checks.

**Comment by CPbN** [comment]
Rather strange, here is my output when building the scripts :
```
# os/linux/snmp
+ centreon/plugins/script_snmp.pm
+ centreon/plugins/snmp.pm
+ centreon/plugins/misc.pm
+ snmp_standard/mode/cpu.pm
+ centreon/plugins/mode.pm
+ centreon/plugins/perfdata.pm
+ snmp_standard/mode/cpudetailed.pm
+ centreon/plugins/statefile.pm
+ snmp_standard/mode/diskio.pm
+ centreon/plugins/templates/counter.pm
+ centreon/plugins/values.pm
+ snmp_standard/mode/diskusage.pm
+ snmp_standard/mode/inodes.pm
+ snmp_standard/mode/interfaces.pm
+ snmp_standard/mode/loadaverage.pm
+ snmp_standard/mode/listdiskspath.pm
+ snmp_standard/mode/listinterfaces.pm
+ snmp_standard/mode/liststorages.pm
+ snmp_standard/mode/memory.pm
+ snmp_standard/mode/processcount.pm
+ snmp_standard/mode/storage.pm
+ snmp_standard/mode/swap.pm
+ snmp_standard/mode/ntp.pm
+ snmp_standard/mode/tcpcon.pm
+ snmp_standard/mode/uptime.pm
+ centreon/plugins/script.pm
+ centreon/plugins/output.pm
+ centreon/plugins/options.pm
+ centreon/plugins/alternative/Getopt.pm
+ centreon/plugins/alternative/FatPackerOptions.pm
```

And :
```
# grep cpu.pm fatpack/os-linux-snmp.pl 
$fatpacked{"snmp_standard/mode/cpu.pm"} = '#line '.(1+__LINE__).' "'.__FILE__."\"\n".<<'SNMP_STANDARD_MODE_CPU';
```

What about you ?

**Comment by joschi99** [comment]
Hmm, on my site i got only this and not more
```
# os/linux/snmp
+ centreon/plugins/script.pm
+ centreon/plugins/output.pm
+ centreon/plugins/misc.pm
+ centreon/plugins/options.pm
+ centreon/plugins/alternative/Getopt.pm
+ centreon/plugins/alternative/FatPackerOptions.pm
+ centreon/plugins/script_snmp.pm
+ centreon/plugins/snmp.pm
```
I'm using centreon plugins version 20180528

**Comment by CPbN** [comment]
Could you give the result of the following command please ?
```
awk -F\' '/^\s*%{ *\$self->{([a-z][a-z]*_)?modes} *} = \(/ {p=1} p && $4 {print} /);/ {p=0}' os/linux/snmp/plugin.pm
```

**Comment by CPbN** [comment]
If it fails, try this one :
```
awk -F\' '/^\s*%{ *\$self->{([a-z][a-z]*_)?modes} *} = \(/ {p=1} p && $4 {print} /\);/ {p=0}' os/linux/snmp/plugin.pm
```

**Comment by CPbN** [comment]
Should be corrected thanks to the last commit.

**Comment by CPbN** [comment]
@Sims24 I think we're OK now, so you could give it a try.
Of course, if you merge it, feel free to squash before !

**Comment by joschi99** [comment]
I have still the issue with missing components. With latest fix always the same output:
```
# os/linux/snmp
+ centreon/plugins/script.pm
+ centreon/plugins/output.pm
+ centreon/plugins/misc.pm
+ centreon/plugins/options.pm
+ centreon/plugins/alternative/Getopt.pm
+ centreon/plugins/alternative/FatPackerOptions.pm
+ centreon/plugins/script_snmp.pm
+ centreon/plugins/snmp.pm
```

**Comment by CPbN** [comment]
@joschi99 quite strange, could you please run the following command :
```
awk -F\' '/^\s*%{ *\$self->{([a-z][a-z]*_)?modes} *} = \(/ {p=1} p && $4 {print} /\);/ {p=0}' os/linux/snmp/plugins.pm`

**Comment by CPbN** [comment]
Mmmh I may have finally managed to reproduce your issue on a server with a different `awk` version.
I'm investigating.

**Comment by joschi99** [comment]
Hi @CPbN,
a first look seem's 👍 and the plugins seem's working very well. Tomorrow I will give you a detailed feedback.

**Comment by joschi99** [comment]
Hi @CPbN,
I tested today different plugins:
* databases: MS SQL, Oracle, MySQL
* VMWare connector
* OS: Windows, Linux
* Network devices: Cisco switches, Cisco WLC, Palo Alto
* UPS: APC
* APPS: SMTP, TCP, DNS

No problems found, you did a great job

**Comment by CPbN** [comment]
Good news, thank you very much for your feedback and your numerous tests @joschi99 :+1:

**Comment by CPbN** [comment]
This time it's then ready, you should be able to safely give it a try @Sims24 :+1:

**Comment by CPbN** [comment]
Did you managed to give this a try @Sims24 ?

**Comment by Sims24** [comment]
Actually in touch with some community members us Raspberry and Debian to get their feedbacks. I'll come back with answers on this soon.

**Comment by CPbN** [comment]
Did you finally get some news @Sims24 ?
Thank you !

**Comment by Sims24** [comment]
Hi, 

Thanks for your work. We implemented ci/cd on our plugins and plugin-packs on our side and refactor some code to generate the fatpacks. 

We will soon provide a link on our official download site allowing everybody to download last fatpacks from all plugins.

Therefore, we will not merge this as it's rather a tool around the plugin and not a check script itself. 

We'll update you as soon as the site offers the link. 

Regards

**Comment by CPbN** [comment]
Good news, thank you very much @Sims24 :+1:

**Comment by CPbN** [comment]
I was thinking about the following, what about a CI solution (Travis CI) which would build the fatpack versions (thanks to this PR script for example) and make them available, everytime a commit to _master_ branch is done ?
Would be a convenient and nice solution :+1:

**Comment by cgagnaire** [comment]
Hi @CPbN,
Don't worry, we have everything we need to do so, like @Sims24 said we'll update asap.
Thanks.

**Comment by CPbN** [comment]
For those still interested into this script, here's an update :
https://github.com/CPbN/centreon-plugins/blob/fatpack/contrib/fatpack_build.sh
I just added support for `mymodule_load` pattern recently added.

**Comment by cgagnaire** [comment]
Here it is at the bottom of Centreon Sources tab : https://download.centreon.com/

**Comment by CPbN** [comment]
Thank you @cgagnaire :+1:

Some comments if I may :
- some plugins are missing in the package ;
- would also be nice to have a nightly build or something like this to be able to benefit from the new developments as soon as possible.

My 2 cents :) Thank you !

**Pull request closed without merging**